experimental/PlaneCurve/EllCurveZnZ.jl

```diff
@@ -212,3 +213,534 @@ function rand_pair_EllCurve_Point(R::Oscar.MPolyRing_dec{S}, PP::Oscar.Geometry.
+    end
+    s1 = Hecke.data(X[2])
+    if !isone(even)
+       !isone(S(D)) && (false, (ZZ(0), ZZ(0)))
+       s2 = sqrtmod(-D, even)
+       s = abs(crt(-s1, odd, -s2+ZZ(2)^(Int(log2(even)-1)), even, true))
+    else
+       s = s1
+    end
+
+    if 2*s > m
+        B = m - s
+    else
+	    B = s
+    end
+
+    A = m
+    while B^2 >= m
+       dr = divrem(A, B)
+       A = B
+       B = dr[2]
+    end
+    t = m - B^2
+    L = divrem(t, D)
+    if !iszero(L[2])
+       return (false, (ZZ(0), ZZ(0)))
+    elseif !issquare(L[1])
+        return (false, (ZZ(0), ZZ(0)))
+    else
+        return (true, [B, isqrt(L[1])])
+    end
+end
+
+################################################################################
+###############Hilbert Class Polynomial#########################################
+################################################################################
+
+#Compute reduced forms of imaginary quadratic number fields
+# based on Algo 5.3.5 in the book by Cohen
+function class_data(D::fmpz)
+   D<ZZ(0) || error("Input needs to be negative")
+
+   h = 0
+   b = abs(D % 2)
+   sqrD3 = sqrt(div(BigInt(-D), 3))
+   B = BigInt(floor(sqrD3))
+   L = Vector{Tuple{fmpz,fmpz,fmpz}}()
+
+   while b <= B
+      q = div(b^2-D, 4)
+      a = ZZ(max(b, 1))
+      while a^2 <= q
+         if q % a == 0 && gcd(a,b, div(q,a)) == 1
+            if a == b || a^2 == q || b == 0
+               h = h+1
+               push!(L, (a,b, div(b^2-D, 4*a)))
+            else
+               h = h + 2
+               push!(L, (a,-b, div(b^2-D, 4*a)))
+               push!(L, (a,b, div(b^2-D, 4*a)))
+
+            end
+         end
+         a = a + 1
+      end
+      b = b + 2
+   end
+   return (ZZ(h), L)
+end
+
+
+function roundarb(x::arb)
+   test = unique_integer(x)
+   test[1] && return test[2]
+   xf = floor(x)
+   x-xf < 0.5 && return xf
+   return xf + 1
+end
+
+function eta_function(tau::acb, n::Int = 17)
+   q = exppii(2*tau)
+   res = sum([(-1)^k*(q^(div(k*(3k-1), 2))+q^(div(k*(3k+1), 2))) for k in 1:n])
+   return (q, 1+res)
+end
+
+function delta_function(tau::acb, n::Int = 17)
+   (q, eta) = eta_function(tau::acb, n)
+   return q*eta^24
+end
+
+function j_function(tau::acb, n::Int = 17)
+   f_tau = inv(delta_function(tau, n))*delta_function(2*tau, n)
+   return inv(f_tau)*(256*f_tau+1)^3
+end
+
+# Computing the Hilbert class polynomial
+function hilbert_class_polynomial(D::fmpz)
+    D>=ZZ(0) && error("Input needs to be negative!")
+    (h, L) = class_data(D)
+    # initial data
+    i = 1
+    sqrD = sqrt(BigInt(-D))
+
+    # estimate bit precision, summand 10 is experimental
+    rp_sum = sum([inv(BigInt(L[i][1])) for i in 1:BigInt(h)])
+    pr_bit = Int(binomial(h, div(h, 2)))*Int(ceil(pi*sqrD*rp_sum*inv(log(2))))+33
+
+    # estimate helper for needed value for sequence
+    pr_seq = inv(3*pi*sqrD)*(2*log(6)+pr_bit*log(2))
+    m = maximum([BigInt(L[i][1]) for i in 1:BigInt(h)])
+    n = Int(ceil(sqrt(pr_seq*m)))
```

Comment:
`floor`, `ceil` and `round` are supposed to be used like `round(Int, bla)`, not `Int(round(bla))` .